src/runtime/profiler.cpp

```diff
@@ -60,13 +76,257 @@ WEAK halide_profiler_pipeline_stats *find_or_create_pipeline(const char *pipelin
         p->funcs[i].stack_peak = 0;
         p->funcs[i].active_threads_numerator = 0;
         p->funcs[i].active_threads_denominator = 0;
+        #ifdef HARDWARE_COUNTERS
+        p->funcs[i].hardware_counters = counters + i;
+        #else
+        p->funcs[i].hardware_counters = NULL;
+        #endif
     }
     s->first_free_id += num_funcs;
     s->pipelines = p;
     return p;
 }
 
-WEAK void bill_func(halide_profiler_state *s, int func_id, uint64_t time, int active_threads) {
+#ifdef HARDWARE_COUNTERS
+WEAK void bill_hardware_counters(halide_profiler_func_stats *f,
+                                 uint64_t active_counters,
+                                 halide_profiler_hardware_counter *before,
+                                 halide_profiler_hardware_counter *after) {
+    if (f->hardware_counters && active_counters) {
+        halide_profiler_hardware_counter *dst = (halide_profiler_hardware_counter *)(f->hardware_counters);
+        halide_profiler_hardware_counter *b = (halide_profiler_hardware_counter *)before;
+        halide_profiler_hardware_counter *a = (halide_profiler_hardware_counter *)after;
+        while (active_counters) {
+            uint64_t i = __builtin_ctzl(active_counters);
```

Comment:
What's the logic here? It looks like iterates over the set bits from LSB to MSB?

If so, is it really worth the clever count trailing zeros, vs. just iterating over all 64 bits? It seems like while the loop might execute fewer iterations, all the loads/stores below would become more predictable (maybe). And it definitely would be easier to read.